### DISCUSSION TITLE
RunTimeError: Factor is exactly singular - fipy 1D diffusion mesh example

### DISCUSSION THREAD

**Discussion by kp2037** [usnistgov/fipy#894] [Q&A] [answered]
I'm new to fipy so I was working through the examples in the documentation (fipy examples page (diffusion.mesh1D) - https://www.ctcms.nist.gov/fipy/examples/diffusion/generated/examples.diffusion.mesh1D.html#module-examples.diffusion.mesh1D).
I keep getting the error "RunTimeError: Factor is exactly singular" when I try to run this piece of code:

```
# diffusion of a hat function
nx = 50 #solution points
dx = 1. #grid spacing
mesh = Grid1D(nx=nx, dx=dx) #grid
L = nx * dx #length of domain
x = mesh.cellCenters[0]

phi = CellVariable(mesh=mesh, name=r"$\phi$") #solution variable
phi.value = 0. #IC
phi.setValue(1., where=(x > L/2. - L/10.) & (x < L/2. + L/10.)) #IC (the hat part)
if __name__ == '__main__':
    viewer = Viewer(vars=phi, datamin=-0.1, datamax=1.1) #plot it

D = 1. #diffusion coefficient
eq = TransientTerm() == DiffusionTerm(D) #pde
dt = 10. * dx**2 / (2 * D) #length of timestep
steps = 100 #number of steps taken

for step in range(steps):
    eq.solve(var=phi, dt=dt) #solver
    viewer.plot()

# now for steady state
phi.value = 0. #reinitialize the solution
phi.setValue(1., where=(x > L/2. - L/10.) & (x < L/2. + L/10.))

DiffusionTerm(coeff=D).solve(var=phi)  #steady state pde
viewer.plot()
```

I can plot the first one fine - the solution with time-dependence, but when I try to reinitialize the solution and solve steady state I get the error I stated above.  Any reasons this might be happening?  All other examples have run perfectly so I don't think it's a problem with the Python version I'm using or Jupyter Lab.

**Comment by guyer** [comment, marked as answer]
Based on the error, you appear to running the SciPy solver suite. I don't know why this error doesn't trigger when we run our automated tests, but the point of this example is that the equation doesn't solve properly. Without boundary conditions, this equation admits an infinite number of solutions. Most solvers conclude that the solution is zero everywhere but, evidently, scipy decides the matrix is singular, but only sometimes.

I get the same error if I run `python examples/diffusion/mesh1D.py --scipy`, but not if I run `python setup.py test --examples --scipy`.
